pkg/admin/data/backup.go

```diff
@@ -0,0 +1,136 @@
+package data
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+// MakeBackup backs up MicroShift data (/var/lib/microshift) to
+// target/name/ (e.g. /var/lib/microshift-backups/backup-00001).
+func makeBackup(cfg BackupConfig) error {
+	if cfg.Storage == "" {
+		return fmt.Errorf("backup storage must not be empty")
+	}
+	if cfg.Name == "" {
+		return fmt.Errorf("backup name must not be empty")
+	}
+
+	if err := microshiftShouldNotRun(); err != nil {
+		return err
+	}
+
+	if err := ensureDirExists(cfg.Storage); err != nil {
+		return err
+	}
+
+	dest := filepath.Join(cfg.Storage, cfg.Name)
+	dest_tmp := dest + ".tmp"
+
+	if err := dirShouldNotExist(dest_tmp); err != nil {
+		return err
+	}
+
+	if err := copyDataDir(dest_tmp); err != nil {
+		return err
+	}
+
+	if err := removeDirBeforeRenaming(dest); err != nil {
+		return err
+	}
+
+	if err := renameDir(dest_tmp, dest); err != nil {
+		return err
+	}
+
+	klog.InfoS("Data backed up", "data", config.DataDir, "backup", dest)
+	return nil
+}
+
+func ensureDirExists(path string) error {
+	klog.Infof("Making sure %s exists", path)
+
+	found, err := util.FileExists(path)
+	if err != nil {
+		return fmt.Errorf("failed checking if %s exists: %w", path, err)
+	}
+	if found {
+		klog.Infof("Directory %s already exists", path)
+		return nil
+	}
+
+	err = util.MakeDir(path)
+	if err != nil {
+		return fmt.Errorf("failed creating %s: %w", path, err)
+	}
+	return err
```

Comment:
It would be good to log here that the directory was created.

I think also explicitly returning nil, instead of err, would make it clearer that this is the success outcome.